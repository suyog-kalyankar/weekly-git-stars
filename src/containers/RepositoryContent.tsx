import { Grid, Link, Typography } from "@mui/material";
import { RepositoryContentProps, StarredParams } from "../types";
import AvtarChip from "../components/AvtarChip";
import { REPO_LINK } from "../constants/AppConstants";
import { getUserReposUrl } from "../util/util";
import {
  BadgeOutlined,
  DescriptionOutlined,
  LanguageOutlined,
} from "@mui/icons-material";
import StarHandler from "../components/Star";

const RepositoryContent = ({
  repository,
  starredRepos,
  setStarredRepoIds,
}: RepositoryContentProps) => {
  const handleStarChange = (starredRepo: StarredParams) => {
    const localStorageRepoIds = localStorage.getItem("repoIds");
    let repositoryIds: number[];
    if (starredRepo.selected) {
      if (
        localStorageRepoIds &&
        localStorageRepoIds !== "undefined" &&
        localStorageRepoIds !== "null"
      ) {
        repositoryIds = [...JSON.parse(localStorageRepoIds), starredRepo.id];
      } else {
        repositoryIds = [starredRepo.id];
      }
    } else {
      repositoryIds =
        localStorageRepoIds &&
        JSON.parse(localStorageRepoIds)?.filter(
          (id: number) => id !== starredRepo.id
        );
    }
    localStorage.setItem("repoIds", JSON.stringify(repositoryIds));
    setStarredRepoIds(repositoryIds);
  };

  return (
    <Grid
      xs={12}
      item
      className="repo-content"
      sx={{
        flexDirection: { xs: "column", md: "row" },
      }}
    >
      <Grid className="initial-content">
        <AvtarChip
          userName={repository?.owner?.login}
          avtarUrl={repository?.owner?.avatar_url}
        />
        <Typography className="center-align" variant="body2">
          <LanguageOutlined /> &nbsp; {repository?.language || "-"}
        </Typography>
      </Grid>
      <Grid className="mid-content">
        <Typography className="center-align" variant="body2">
          <BadgeOutlined /> &nbsp; {repository?.name || "-"}
        </Typography>
        <Typography className="center-align" variant="body2">
          <DescriptionOutlined /> &nbsp;{" "}
          {repository?.description || "No description available"}
        </Typography>
        <Link
          href={getUserReposUrl(repository.owner.login, repository?.name)}
          target="_blank"
          sx={{ color: "midnightblue" }}
        >
          {REPO_LINK}
        </Link>
      </Grid>
      <StarHandler
        selectedId={repository.id}
        onStarredChange={handleStarChange}
        starred={starredRepos?.includes(repository.id) ? true : false}
      />
    </Grid>
  );
};

export default RepositoryContent;
