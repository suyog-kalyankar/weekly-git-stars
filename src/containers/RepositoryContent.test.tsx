import React from "react";
import { render, screen } from "@testing-library/react";
import RepositoryContent from "./RepositoryContent";

describe("RepositoryContent Component", () => {
  const repositoryMock = {
    id: 1,
    owner: {
      id: 25,
      login: "suyog",
      avatar_url: "https://github-repo.com/avatar.png",
    },
    language: "JavaScript",
    name: "test-repo",
    description: "This is a JavaScript repository",
  };

  const starredReposMock = [1, 2, 3];

  const setStarredRepoIdsMock = jest.fn();

  it("renders without crashing", () => {
    const { container } = render(
      <RepositoryContent
        key={repositoryMock.id}
        repository={repositoryMock}
        starredRepos={starredReposMock}
        setStarredRepoIds={setStarredRepoIdsMock}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("displays repository information correctly", () => {
    render(
      <RepositoryContent
        repository={repositoryMock}
        key={repositoryMock.id}
        starredRepos={starredReposMock}
        setStarredRepoIds={setStarredRepoIdsMock}
      />
    );

    expect(screen.getByText(repositoryMock.language)).toBeInTheDocument();
    expect(screen.getByText(repositoryMock.name)).toBeInTheDocument();
    expect(screen.getByText(repositoryMock.description)).toBeInTheDocument();
    expect(screen.getByRole("link")).toHaveAttribute(
      "href",
      expect.stringContaining(repositoryMock.owner.login)
    );
  });
});
