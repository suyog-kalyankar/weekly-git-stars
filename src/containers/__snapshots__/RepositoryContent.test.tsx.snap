// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryContent Component renders without crashing 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 repo-content css-15a6vvq-MuiGrid-root"
  >
    <div
      class="MuiGrid-root initial-content css-vj1n65-MuiGrid-root"
    >
      <div
        class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault chip css-gavykb-MuiChip-root"
        data-testid="avtar-chip"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiChip-avatar MuiChip-avatarMedium MuiChip-avatarColorDefault css-1wlk0hk-MuiAvatar-root"
        >
          <img
            alt="avtar-img"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://github-repo.com/avatar.png"
          />
        </div>
        <span
          class="MuiChip-label MuiChip-labelMedium css-6od3lo-MuiChip-label"
        >
          suyog
        </span>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2 center-align css-e784if-MuiTypography-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="LanguageOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2m6.93 6h-2.95c-.32-1.25-.78-2.45-1.38-3.56 1.84.63 3.37 1.91 4.33 3.56M12 4.04c.83 1.2 1.48 2.53 1.91 3.96h-3.82c.43-1.43 1.08-2.76 1.91-3.96M4.26 14C4.1 13.36 4 12.69 4 12s.1-1.36.26-2h3.38c-.08.66-.14 1.32-.14 2s.06 1.34.14 2zm.82 2h2.95c.32 1.25.78 2.45 1.38 3.56-1.84-.63-3.37-1.9-4.33-3.56m2.95-8H5.08c.96-1.66 2.49-2.93 4.33-3.56C8.81 5.55 8.35 6.75 8.03 8M12 19.96c-.83-1.2-1.48-2.53-1.91-3.96h3.82c-.43 1.43-1.08 2.76-1.91 3.96M14.34 14H9.66c-.09-.66-.16-1.32-.16-2s.07-1.35.16-2h4.68c.09.65.16 1.32.16 2s-.07 1.34-.16 2m.25 5.56c.6-1.11 1.06-2.31 1.38-3.56h2.95c-.96 1.65-2.49 2.93-4.33 3.56M16.36 14c.08-.66.14-1.32.14-2s-.06-1.34-.14-2h3.38c.16.64.26 1.31.26 2s-.1 1.36-.26 2z"
          />
        </svg>
           
        JavaScript
      </p>
    </div>
    <div
      class="MuiGrid-root mid-content css-vj1n65-MuiGrid-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 center-align css-e784if-MuiTypography-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="BadgeOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14 12h4v1.5h-4zm0 3h4v1.5h-4z"
          />
          <path
            d="M20 7h-5V4c0-1.1-.9-2-2-2h-2c-1.1 0-2 .9-2 2v3H4c-1.1 0-2 .9-2 2v11c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V9c0-1.1-.9-2-2-2m-9 0V4h2v5h-2zm9 13H4V9h5c0 1.1.9 2 2 2h2c1.1 0 2-.9 2-2h5z"
          />
          <circle
            cx="9"
            cy="13.5"
            r="1.5"
          />
          <path
            d="M11.08 16.18c-.64-.28-1.34-.43-2.08-.43s-1.44.15-2.08.43c-.56.24-.92.78-.92 1.39V18h6v-.43c0-.61-.36-1.15-.92-1.39"
          />
        </svg>
           
        test-repo
      </p>
      <p
        class="MuiTypography-root MuiTypography-body2 center-align css-e784if-MuiTypography-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DescriptionOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M8 16h8v2H8zm0-4h8v2H8zm6-10H6c-1.1 0-2 .9-2 2v16c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8zm4 18H6V4h7v5h5z"
          />
        </svg>
          
         
        This is a JavaScript repository
      </p>
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-17f922i-MuiTypography-root-MuiLink-root"
        href="https://github.com/suyog/test-repo"
        target="_blank"
      >
        Visit repository
      </a>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium star-btn css-sghohy-MuiButtonBase-root-MuiButton-root"
      role="star-btn-role"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium star-icon css-i4bv87-MuiSvgIcon-root"
          data-testid="StarIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
          />
        </svg>
      </span>
      Starred
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
